### ISSUE TITLE
Template Lifecycle Management (goals, desired UX specification, etc.)

### ISSUE THREAD

**Issue by DinaBelova** [Mirantis/hmc#296]
**Goals**
Deliver a viable solution to effectively manage the life cycle of templates that will work for both the k8s cluster and the state management.
Support independent lifecycle management of templates from the 2A core systems whilst allowing for referencing of dependencies.

Templates will be immutable and for each Helm Chart Version of a Template a new Template object needs to be created, this Epic will cover how a Platform Lead can define which templates should actually be usable by the Platform Engineers and how they can be used.

**Major deliverables**
- Standard system to maintain the template lifecycle regardless of the source of the template
- Templates can be version controlled with a clear way to show which is the latest available template in the chain
- Kubernetes native way to distribute and control access to templates 
- Cluster operators can choose when a cluster is updated

Upgrade Principles
- Only N-3 Template Versions need to be supported by Mirantis (Supported = Mirantis provides fixes and updates to only these versions and not older )
- Templates updates must be iterative from the previous template (this is an assumption: the person creating the templates will understand that massive changes in templates will cause problems, 2A does not need to validate the amount of changes for now, this might be done in the future)
- Templates should be immutable and always require a new template to be created 
- Software version changes e.g. k8s updates will require a new template
- Template creation and delivery process should be the same regardless of the source of the template

**Who it benefits**
-  Customer: Repeatable deployments defined by Platform engineering leaders that ensure consistent clusters regardless of who deploys them and where they are deployed
- Platform Engineering Teams: Simplified deployment of clusters and dependencies as well as lifecycle management of the clusters

**Acceptance criteria**
New Template
- Platform lead can create a new template and associated artifacts (TemplateChain Objects, etc)
- New templates can be associated to a specific release of 2A that is supported 

Template Updates
- Platform lead can provide a updated a template and make it available 
- Platform engineer is able to easily see when a updated Template is available for a cluster
- Platform engineer can choose when a cluster is updated
- Platform Engineer can select from the available updates for a cluster
- Update dependencies are validated before updating 
- Updates to templates will not impact a running cluster unless specifically approved by the Platform Engineer (cluster owner)

Template Access
- Platform lead can assign - and - remove a template from a particular user
- New Templates created by Mirantis or the Platform Lead cannot be used by Platform Engineer unless they are made available by the Platform Lead. 
- Templates that have been created but not yet approved can be used by Platform Lead for testing

**Assumptions**
- Templates (helm charts and CR) are immutable and require a new one to be created
- Multiple repository sources will be supports within 2A
- The actual updates will be managed by 2A (HMC + CAPI)

**Out of scope**
This epic will cover only the update of the templates and access control for the templates, validating the deployed cluster is out of scope for now.

**User stories**
- As a Platform engineer I want to be able to upgrade my kubernetes clusters created from a template in a consistent manner
- As a platform engineer I want to be able to select when my kubernetes upgrade happens.
- As a platform lead I want to to be able to pin my kubernetes cluster to a specific kubernetes versions
- As a platform engineer I want to be able to verify that cluster upgrade won’t break any dependencies
- As a platform engineer I want to see the status of a upgrade in one place and be able to troubleshoot why a deployment has failed

**Related Implementation GH issues**
- #267
- #266
- #317 (depends on #266)

**Comment by DinaBelova** [comment]
only the final minor task #329 remains in the related epic #266 -> I think we may close this epic.
